Make VS Project system components free threaded

mavasani opened the pull request:
1. DocumentProvider: DocumentProvider can now serve requests to open/close documents or register opened documents with project from background threads. Any operations that need foreground thread access (such as accessing RDT, VS Hierarchy, etc.) are scheduled on separate per-document tasks executed with the foreground task scheduler. Accesses to data structures that might be accessed from multiple threads are protected with locks.

2. VisualStudioProjectTracker: VisualStudioProjectTracker can now serve requests to add/remove projects, update binary paths used to track projects or notify workspace hosts of project state changes from background threads. Any operations that need foreground thread access are scheduled on the foreground task scheduler. Accesses to data structures that might be accessed from multiple threads are protected with locks.

3. AbstractProject: Similar to above, certain project system operations such add/remove of documents/references/analyzers, updating options, etc. can now be invoked from background threads.

mavasani commented:
test vsi please

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -65,6 +55,23 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
+        /// change comes in before we process it.
+        /// </summary>
+        private readonly Dictionary<VisualStudioMetadataReference, CancellationTokenSource> _changedReferencesPendingUpdate
+            = new Dictionary<VisualStudioMetadataReference, CancellationTokenSource>();
```
Consider: 

1. Ban all uses of the field directly 
2. Add a `private` property to access the field
3. The `private` property `get` should assert it is on the foreground thread

mavasani replied:
Sure.

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -145,27 +140,87 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
         public bool IsWebSite { get; protected set; }
 
+        #region ObjOutputPath
+        private string _objOutputPath;
```
New field that doesn't have a comment about threading policy.

mavasani replied:
I'll add comments and clean up further, sent this PR out early to know if I had broken something real bad.

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -145,27 +140,87 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
+        #endregion
 
+        #region BinOutputPath
+        private string _binOutputPath;
```
New mutable field that doesn't have a comment about threading policy.

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -145,27 +140,87 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
 
-        public IRuleSetFile RuleSetFile { get; private set; }
+        #region RuleSetFile
+        private IRuleSetFile _ruleSetFile;
```
Looks like you added comments about threading policy to all existing fields, but no comments for new fields.  Please make sure all fields have their threading policy addressed.

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -217,19 +290,80 @@ protected string ContainingDirectoryPathOpt
+                lock (_gate)
+                {
+                    _assemblyName = value;
+                }
```
I question the value of the majority of locks you added in property implementations.  Can you detail why they are needed?  

The values here are all reference types hence we already know the assignments are atomic.  The presence of `lock` is only useful when you need to synchronize those read / writes with other operations.  I don't see that happening here.

mavasani replied:
The primary reason I though about adding a lock in the setters is to ensure that we don't allow writes to any fields while creating ProjectInfo from current state [here](https://github.com/dotnet/roslyn/pull/12927/files#diff-c1adde957bdde82f66d830cea57b2ddfR434). However, looking at the code invoking [CreateProjectInfoForCurrentState](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioProjectTracker.WorkspaceHostState.cs,90), we probably need to prevent any state changes (or queue them) between invocation of [CreateProjectInfoForCurrentState](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/fe613ec5e45c2f82.html) and [StartPushingToWorkspaceHosts](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/922222fcc2607de1.html), and subsequent changes will anyway be pushed to the workspace hosts. Need to think a bit more on this.

mavasani commented:
test vsi please

mavasani commented:
test vsi please

mavasani commented:
Tagging @jasonmalinowski @dotnet/roslyn-ide @dotnet/project-system for review.

jaredpar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -560,9 +709,16 @@ private void OnImportChangedAfterDelay(Task previous, object state)
-            _changedReferencesPendingUpdate.Remove(reference);
+            ChangedReferencesPendingUpdate.Remove(reference);
+
+            bool hasMetadataReference;
```
> bool hasMetadataReference; [](start = 11, length = 27)

Nit: should be called hadMetadataReference

mavasani commented:
retest vsi please

mavasani commented:
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Retest reason: https://github.com/dotnet/roslyn/issues/12651#issuecomment-239956994

mavasani commented:
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Retest reason: https://github.com/dotnet/roslyn/issues/13189

mavasani commented:
Ping?

mavasani commented:
Tagging @jasonmalinowski @Pilchie @dotnet/roslyn-ide for review.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -94,8 +118,49 @@ public VisualStudioProjectTracker(IServiceProvider serviceProvider)
+
+            lock (_gate)
+            {
+                _taskForForegroundAffinitizedActions.Wait();
```
How does this work? If we're running this on the UI thread, and these tasks are scheduled to the UI thread, why won't this deadlock?

mavasani replied:
I haven't tested it, most probably it doesn't. I am actually not even sure we need to wait for all scheduled foreground actions here: https://github.com/dotnet/roslyn/pull/12927/files/55b4181c4040cae49c209fc23dd9b8e829cefb55#diff-5561564909d81f97c70a52c751280162R225. These scheduled tasks may add/remove projects, but that shouldn't be no different then these events coming after we have started pushing to workspace host. I am tending towards just removing this, what do you suggest?

jasonmalinowski replied:
I admit I didn't understand why this mechanism existed at all. If you think it's unnecessary then let's delete it, but clearly you thought it was worth adding in the first place!

mavasani replied:
Lets say, I was as much confused as skeptical. If it turns out we need to force complete pending tasks, we will find out later and then fix it.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -189,32 +280,61 @@ public void Dispose()
         /// <summary>
         /// Add a project to the workspace.
+        /// If invoked on the foreground thread, the add is executed right away.
+        /// Otherwise, the add is scheduled on foreground task scheduler.
```
Why is this scheduled for later? Shouldn't we be pushing it to the workspace asynchronously and then dealing with just the open file handling on the UI thread?

mavasani replied:
I have attempted to do the minimum work to 1. avoid switching back to the foreground thread when any free threaded method is invoked on the tracker from background and 2. enable multithreaded access to its data structures - I agree we could do lot more work asynchronously, but that could come in future.

jasonmalinowski replied:
I guess my question is more "why can't I just add projects on a background thread? why are we moving to the foreground thread to notify workspace hosts?"

mavasani replied:
UpdateProjectBinPath_Foreground is making calls back into AbstractProject instances for metadata to P2P reference conversion, and I didn't know what may go wrong if that happens from multiple background threads - I just retained that functionality on foreground thread.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
         internal bool TryGetProjectByBinPath(string filePath, out AbstractProject project)
         {
-            project = null;
+            lock (_projectsByBinPath)
```
Shouldn't this be lock (_gate)? Even if not, I'd much prefer a single gate be used for everything instead of separate gates for separate items.

mavasani replied:
I can switch to a single gate.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
+        /// <remarks>This method may be called on a background thread.</remarks>
+        internal bool TryGetProjectsByBinPath(string filePath, out ImmutableArray<AbstractProject> projects)
+        {
+            lock (_projectsByBinPath)
```
Ditto, lock (_gate)?

mavasani replied:
Actually, I did add a common `_gate` object but just forgot to switch few lock statements to use it - I'll update the PR.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
-            return false;
+        internal void AddProjectByBinPath(string filePath, AbstractProject project)
+        {
+            lock (_projectsByBinPath)
```
Ditto, lock (_gate)?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs:
```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
+
+        internal void RemoveProjectByBinPath(string filePath, AbstractProject project)
+        {
+            lock (_projectsByBinPath)
```
Ditto, lock (_gate)?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -263,28 +269,59 @@ public ProjectInfo CreateProjectInfoForCurrentState()
+                hasAllInformation = this.LastDesignTimeBuildSucceeded;
+            }
+
             var info = ProjectInfo.Create(
```
This call is cheap. It might be more sane to read if we just keep all of this under the lock.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -560,9 +709,16 @@ private void OnImportChangedAfterDelay(Task previous, object state)
+            lock (_gate)
+            {
+                hasMetadataReference = _metadataReferences.Contains(reference);
+            }
```
I'd prefer if we just took the lock once in this function. Otherwise, nothing prevents this answer from changing from somebody else coming through as @jaredpar astutely observes. If/when we see performance problems, let's optimize this. Otherwise this is just asking for subtle races.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -932,37 +1077,50 @@ public virtual void Disconnect()
+                ImmutableArray<VisualStudioMetadataReference> metadataReferences;
+                ImmutableArray<VisualStudioAnalyzer> analyzers;
+
+                lock (_gate)
```
Any reason we can't just take a single gate around this whole thing? Because otherwise you're just opening yourself up to more races.

mavasani replied:
I wasn't sure about the perf for the methods below, especially if the number of docs/references/analyzers is large, so avoided taking the lock for the whole thing. Probably doesn't matter as the project system is anyway disposing this project, I can switch to doing this entirely under the lock.

jasonmalinowski replied:
Multithreaded code isn't a place for premature optimizations!

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -23,14 +25,15 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
         protected readonly IContentTypeRegistryService ContentTypeRegistryService;
 
         private readonly uint _runningDocumentTableEventCookie;
+        private readonly object _gate = new object();
```
Can you comment which fields this covers? I like how you did it for the other types.

mavasani replied:
Yes, let me add more comments and separate out the fields into regions.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -81,58 +85,77 @@ internal abstract partial class DocumentProvider
+                        return null;
+                    }
                 }
             }
```
Shouldn't there be the "else queue the foreground work" thing here?

mavasani replied:
That is actually not needed, I think. I have a comment here: https://github.com/dotnet/roslyn/pull/12927/files/55b4181c4040cae49c209fc23dd9b8e829cefb55#diff-8ab92731d875a419c8808cd50f634912R92

Let me know if I am wrong here.

jasonmalinowski replied:
I guess I'd move that comment closer to the check then.

jasonmalinowski replied:
But I think it makes sense, perhaps just more comments in more places.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+            {
+                // Add taskAndTokenSource to the pending document initialization tasks.
+                // Check for cancellation before adding as the task might already have been completed/cancelled/faulted before we reached here.
+                if (!cts.IsCancellationRequested && !task.IsCompleted && !task.IsFaulted)
```
task.IsCompleted means the task ran to any state, so that includes IsFaulted.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+                // Check for cancellation before adding as the task might already have been completed/cancelled/faulted before we reached here.
+                if (!cts.IsCancellationRequested && !task.IsCompleted && !task.IsFaulted)
+                {
+                    _pendingDocumentInitializationTasks.Add(document.Id, taskAndTokenSource);
```
We recently had some long discussions in the IDE team room about the reusability of Ids across documents. Should you just use the document object itself there?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+            TaskAndTokenSource taskAndTokenSource;
+            if (_pendingDocumentInitializationTasks.TryGetValue(document.Id, out taskAndTokenSource))
+            {
+                // Signal cancellation and wait for task to complete.
```
Why wait? Either the task is already running and it's too late or it hasn't started yet and it's a no-op. I guess the guarantee you're trying to provide isn't clear.

mavasani replied:
I wasn't sure if running task might access data structures with the assumption of the document still being active, while the callee doing the cancellation might proceed after this invocation and remove the document entry from the data structures.

jasonmalinowski replied:
I'd make it simpler by just taking a lock for both the removal side and the processing of the task side. If you only cancel while the lock is held, and check the cancellation token after taking the lock, then you can prevent that and it might be a lot simpler and clearer.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
```
Comment as to why this is needed for correctnesss. (I suspect I know the answer, which is why it should be commented. :-))

mavasani replied:
Yes, sure.

jasonmalinowski replied:
Or actually: I'd consider that this function shouldn't take a cancellation token. This mutates state, so by the contract it's not clear what cancellation means -- can you end up in some torn state? Did it happen? Did it not happen? But maybe that results in harder to write code, I'm not sure.

mavasani replied:
I had to carefully place the `cancellationToken.ThrowIfCancellationRequested();` calls in this method to avoid corrupting consistency between data structures and I avoided putting it within the loops and I am still not sure if I have done the right thing. Probably better to take away the cancellation token for simplicity and we can add it later if perf necessitates it.

jasonmalinowski replied:
Yep, your concern is also my concern. I'd imagine the only times the task gets cancelled are:

1. We remove the document from the background thread while we're looking for it on the UI thread.
2. We had some UI notification (an RDT event, for example) that meant that the work we queued to the UI thread is no longer necessary.

The second doesn't benefit from more cancellations, since both pieces are on the UI thread. And the first seems rare. I wouldn't worry about optimizing it, especially given the risks you called out.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
+            try
             {
-                foreach (var project in _projectContainer.GetProjects())
+                projectsBuilder.AddRange(_projectContainer.GetProjects());
```
Why are we taking these and then stuffing them in the array builder? Aren't they an ImmutableArray we can enumerate over?

mavasani replied:
Interface defines them to return an IEnumerable: http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/IVisualStudioHostProjectContainer.cs,9a312d9c36dac361,references but the underlying value is now ImmutableArray.

I'll change the interface to return an immutable array and clean this up.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -300,6 +458,16 @@ protected virtual void OnBeforeNonRoslynDocumentWindowShow(IVsWindowFrame frame,
+
+        private IList<DocumentId> GetDocumentIdsFromDocCookie_NoLock(uint docCookie)
+        {
+            AssertIsForeground();
```
Why a _NoLock version that asserts foreground affinity? It's effectively taking the "foreground lock" if you will.

mavasani replied:
Seems better pattern would be to move this assert outside the lock at start of GetDocumentIdsFromDocCookie

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -346,6 +527,7 @@ private void CloseDocuments(uint docCookie, string monikerToKeep)
             // last documentKey in this list.
             var updateActiveContext = documentsToClose.Count == 1;
 
+            var documentBuilder = ArrayBuilder<StandardTextDocument>.GetInstance();
```
Unused?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+            }
+        }
+
+        private void RemovePendingDocumentInitializationTask(IVisualStudioHostDocument document)
```
I don't understand why there's both a RemovePending and CancelPending methods, and why I'd use one or the other. I'd assume the only operation was to cancel, because we either know the document isn't needed anymore (it was closed) or we've already taken care of the operation on the UI thread.

mavasani replied:
Yes, the difference was just the Task.Wait() in the cancel implementation above. Given we don't need it, both of this can be the same method.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -381,24 +563,39 @@ private void OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, IVsHierar
+
+                if (currDocHier == pHierNew)
+                {
+                    var documentKey = documentKeys[i];
```
Consider doing all of this under the gate and then just moving the notification out. It might be easier to reason about. (It took be about a minute to figure out what's up with all the copying of state around.)

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs:
```diff
@@ -172,7 +176,7 @@ protected sealed override void ValidateReferences()
         {
             // can happen when project is unloaded and reloaded or in Venus (aspx) case
-            if (ProjectFilePath == null || TryGetBinOutputPath() == null || TryGetObjOutputPath() == null)
+            if (ProjectFilePath == null || BinOutputPath == null || ObjOutputPath == null)
```
Why the rename here? I think we had the method so it was clear you might not get something back from these.

mavasani replied:
Because the method was just returning a field and wasn't correctly following the TryGet pattern. I also got a request for the same from @davkean. We either need to change this to return bool and take an out param OR change it to a property with a comment about it being possibly null. I took the later approach as TryGet to just return a field seemed weird.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.WorkspaceHostState.cs:
```diff
@@ -90,6 +105,14 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
 
                 var projectInfos = inOrderToPush.Select(p => p.CreateProjectInfoForCurrentState()).ToImmutableArray();
 
+                // Enable projects to start pushing changes to workspace hosts, even before we add the solution/project to the host.
```
Can this comment be expanded? I don't follow it.

mavasani replied:
The issue is that between the point we capture the project info for current state above and the point where we start pushing to workspace hosts, CPS may send new events on the AbstractProject on a background thread, and these won't get pushed over to the workspace hosts as we didn't set the `_pushingChangesToWorkspaceHost` flag on the AbstractProject. If we set this flag upfront, then such background thread events with ensure that notification to workspace host is queued up on foreground scheduled tasks.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs:
```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+        /// If we are on a foregroud thread, then this is done right away.
+        /// Otherwise, we schedule a task on foreground task scheduler.
         /// </summary>
         public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document)
```
Maybe rename this to NotifyDocumentRegisteredToProjectAndStartToRaiseEvents or something? I don't mind it's doing two things, but the contract is clearer.

jasonmalinowski commented:
Overall comments:

1. Not bad! I very much appreciate your clear documenting of all the fields and what threads you expect are touching them and what locks need to be taken. That basic stuff is missing from a lot of multithreaded code out there, so thanks for doing that.
2. I'd favor "big fat locks" for functions more than I think you are doing. There's a few cases where you take a lock, grab a few bits of data, then release the lock, do something, and then take the lock again in the function a bit later. Those may be entirely safe, but the proof of that is a lot harder than just locking around the whole thing. If we see particular sources of contention we can update that, but until then it might be better to keep it simple.
3. I see the trees really well; the forest, less so. I'm not sure if it's good just writing up a markdown file in the docs or something to explain the high-level patterns being employed here, or just comment more heavily. Specifically, the project tracker's use of a queue to notify hosts on the UI thread was unclear to me why that was necessary, and what invariants I'd have to maintain if I was in this corner of the code.

jasonmalinowski commented:
And oh, I assume you're squash-merging this, and if you are make sure to write some proper text in the PR description before merging.

mavasani commented:
test vsi please

mavasani commented:
@jasonmalinowski Thanks for the review. I have updated the PR addressing your feedback. I am merging this on so we can start playing around with background thread initialization in CPS. I'll address any further feedback with a follow up change.
